From f4f50353ad6afca7990b2c2037adc90e5cf10c3c Mon Sep 17 00:00:00 2001
From: "Guoniu.Zhou" <guoniu.zhou@nxp.com>
Date: Fri, 23 Mar 2018 11:41:49 +0800
Subject: [PATCH 3529/5242] MLK-17487: pxp: fix pxp yuv to yuv generate color
 dots issue

commit  dceff7dc1339be044c1588e32807369b8277b6ce from
https://source.codeaurora.org/external/imx/linux-imx.git

When PxP convert yuyv to nv12 format, some color dots will
introdue to output image. IC recommend that YCBCR_MODE and
BYPASS bit of CSC1_COEF0 should be 1.

Reviewed-by: robby.cai <robby.cai@nxp.com>
Signed-off-by: Guoniu.Zhou <guoniu.zhou@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/dma/pxp/pxp_dma_v3.c |    9 ++++++++-
 1 file changed, 8 insertions(+), 1 deletion(-)

diff --git a/drivers/dma/pxp/pxp_dma_v3.c b/drivers/dma/pxp/pxp_dma_v3.c
index 29105c9..b470dba 100644
--- a/drivers/dma/pxp/pxp_dma_v3.c
+++ b/drivers/dma/pxp/pxp_dma_v3.c
@@ -3021,7 +3021,6 @@ static int pxp_2d_task_config(struct pxp_pixmap *input,
 {
 	uint8_t position = 0;
 
-
 	do {
 		position = find_next_bit((unsigned long *)&nodes_used, 32, position);
 		if (position >= sizeof(uint32_t) * 8)
@@ -3151,6 +3150,7 @@ static int pxp_2d_op_handler(struct pxps *pxp)
 	uint32_t partial_nodes_used = 0;
 	uint32_t nodes_used_s0 = 0, nodes_used_s1 = 0;
 	uint32_t nodes_in_path_s0, nodes_in_path_s1;
+	uint32_t val;
 
 	output = &task->output[0];
 	if (!output->pitch)
@@ -3287,6 +3287,13 @@ static int pxp_2d_op_handler(struct pxps *pxp)
 		pr_debug("%s: path_ctrl0 = 0x%x\n",
 			 __func__, *(uint32_t *)&path_ctrl0);
 		pxp_2d_task_config(input, output, op, nodes_used);
+
+		if (is_yuv(input->format) && is_yuv(output->format)) {
+			val = readl(pxp_reg_base + HW_PXP_CSC1_COEF0);
+			val |= (BF_PXP_CSC1_COEF0_YCBCR_MODE(1) |
+					BF_PXP_CSC1_COEF0_BYPASS(1));
+			pxp_writel(val, HW_PXP_CSC1_COEF0);
+		}
 		break;
 	case 2:
 		/* Composite */
-- 
1.7.9.5

